### ISSUE TITLE
Explicitly putting bot name within regex

### ISSUE THREAD

**Issue by novito** [slack-ruby/slack-ruby-bot#127]
I've been trying to find out a way that I can use regex to parse a request to my bot, making sure that I am forced to ping the bot. Example:

Say I have this matcher in my bot:

```
match /^what's the solution to (?<problem>\w*)\?$/i do |client, data, match|
...
end
```

This is fine, but I don't want my bot to answer unless it's explicitly pinged. So I've tried doing this approach: https://github.com/dblock/slack-google-bot/blob/master/slack-google-bot/commands/google.rb#L6

```
match /^(?<bot>\w*)\swhat's the solution to (?<problem>\w*)\?$/i do |client, data, match|
...
end
```

And then messaging in slack like: `my-bot what's the solution to problem1?` without the bot parsing that request.  

What am I missing?

**Comment by dblock** [comment]
You closed this, care to elaborate for the next person?

**Comment by novito** [comment]
@dblock absolutely. Instead of doing:
```match /^(?<bot>\w*)\swhat's the solution to (?<problem>\w*)\?$/i do |client, data, match|```
I did:
```match /^(?<bot>[[:alnum:][:punct:]@<>]*)\swhat's the solution to (?<problem>\w*)\?$/i do |client, data, match|```

I took it from https://github.com/slack-ruby/slack-ruby-bot/blob/master/lib/slack-ruby-bot/commands/base.rb#L49

What I don't understand is how that first part of the regex ```(?<bot>[[:alnum:][:punct:]@<>]*)``` just matches the bot name. It seems to me I would be able to place any random name and it would still match, and I haven't been able to find how the code deals with that. @dblock care to explain?

**Comment by estiens** [comment]
It would be nice to know how this works and actually to abstract it up. 

Something like `command match` to go along with `match` and `command` -- will dig in a bit and see if I can make that PR

**Comment by kstole** [comment]
@estiens I agree with you on this. I've had a similar use case where I basically want to use the `command` method so that it will only match stuff sent to the bot but at the same time, it takes a regex. My solution was to create a custom bot-matching class and send it to the `match` method.
```ruby
class BotRegex < Regexp
  def initialize(*values)
    values = values.map { |value| Regexp.new(value).source }.join('|')
    super("^(?<bot>[[:alnum:][:punct:]@<>]*)[\\s]+(?<command>#{values})([\\s]+(?<expression>.*)|)$", Regexp::IGNORECASE)
  end
end
```
```ruby
match BotRegex.new(/Some\s*regex+\?*/) do |client, data, match|
```
I think adding a new method for this would work well, but I think I'd prefer modifying the `command` method to accept both String and RegExp arguments, and then only escape the String arguments.
```ruby
command /Some\s*regex+\?*/ do |client, data, match|
```

**Comment by dblock** [comment]
Great idea @kstole!
